### PR TITLE
[enhancement](nereids)only push having as agg's parent if having just use slots from agg's output

### PR DESCRIPTION
**Pull request by starocean999**
## Proposed changes

1. only push having as agg's parent if having just use slots from agg's output
2. show user friendly error message when item in select list but not in aggregate node's output

## Further comments

If this is a relatively large or complex change, kick off the discussion at [dev@doris.apache.org](mailto:dev@doris.apache.org) by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

**Comment by starocean999** [comment]
run buildall

**Comment by starocean999** [comment]
run buildall

**Comment by starocean999** [comment]
run buildall

**Comment by starocean999** [comment]
run buildall

File: fe/fe-core/src/main/java/org/apache/doris/nereids/rules/analysis/NormalizeAggregate.java [hunk 1]
```diff
@@ -244,6 +246,20 @@ private LogicalPlan normalizeAgg(LogicalAggregate<Plan> aggregate, Optional<Logi
...
         LogicalProject<Plan> project = new LogicalProject<>(upperProjects, newAggregate);
+        // verify project used slots are all coming from agg's output
+        List<Slot> slots = collectAllUsedSlots(upperProjects);
+        if (!slots.isEmpty()) {
+            Set<ExprId> aggOutput = normalizedAggOutput.stream()
+                    .map(NamedExpression::getExprId)
+                    .collect(Collectors.toSet());
+            slots = slots.stream()
+                    .filter(s -> !aggOutput.contains(s.getExprId()))
+                    .collect(Collectors.toList());
```
Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/rules/analysis/NormalizeAggregate.java:257]: use for loop for better perfermance

File: fe/fe-core/src/main/java/org/apache/doris/nereids/rules/analysis/NormalizeAggregate.java [hunk 2]
```diff
@@ -287,4 +303,11 @@ private List<NamedExpression> normalizeOutput(List<NamedExpression> aggregateOut
         }
         return builder.build();
     }
+
+    private List<Slot> collectAllUsedSlots(List<NamedExpression> expressions) {
+        return Stream.concat(
+                ExpressionUtils.collectAll(expressions, SubqueryExpr.class::isInstance).stream()
```
Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/rules/analysis/NormalizeAggregate.java:309]: ditto

**Comment by starocean999** [comment]
run buildall

**Comment by starocean999** [comment]
run buildall

**Review by morrySnow — DISMISSED** [review]

**Comment by starocean999** [comment]
run buildall

**Review by morrySnow — APPROVED** [review]

**Review by 924060929 — APPROVED** [review]

**Pull request merged**